src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala

```diff
@@ -129,75 +148,79 @@ object RockSlicer {
-    *         matching processor joints. Redundant faces and duplicate blocks are already removed
-    *         in this list, so it is not necessary to perform these computations on the returned
-    *         sequence.
+    *                     input is an empty sequence
+    * @param orphanBlocks Sequence of blocks that cannot be matched with other blocks. Initial
+    *                     input is an empty sequence
+    * @return Tuple containing two sequences of blocks. The first sequence is all the merged blocks.
+    *         The second is all blocks that could not be merged.
     */
   @tailrec
   def mergeBlocks(processorBlocks: Seq[Block], mergedBlocks: Seq[Block],
-                  origin: (Double, Double, Double)): Seq[Block] = {
-    val reconstructedBlocks = (processorBlocks.tail map { block =>
-      val currentBlock = Block(origin, processorBlocks.head.updateFaces(origin))
-      val updatedBlock = Block(origin, block.updateFaces(origin))
-      val sharedProcFaces = compareProcessorBlocks(currentBlock, updatedBlock)
-      if (sharedProcFaces.nonEmpty) {
-        val currentFaces = currentBlock.faces.diff(sharedProcFaces)
-        val updatedFaces = updatedBlock.faces.diff(sharedProcFaces)
-        val allFaces = currentFaces ++ updatedFaces
-        // Check for any actual shared faces between blocks - if these exists blocks
-        // should not be merged since there is a real joint seperating the blocks
-        val nonSharedFaces =
-          allFaces.foldLeft(Seq.empty[Face]) { (unique, current) =>
-            if (!unique.exists(current.isSharedWith(_))) {
-              current +: unique
-            } else {
-              unique
-            }
-          }
-        if (allFaces.diff(nonSharedFaces).isEmpty) {
-          Some(Block(origin, nonSharedFaces))
-        } else {
-          None
-        }
+                  orphanBlocks: Seq[Block]):
+                 (Seq[Block], Seq[Block]) = {
+    val blockMatches = findMates(processorBlocks)
+    val joinedBlocks = blockMatches.map{ case (paired, _) => paired }
+    val matchIndices = blockMatches.flatMap { case (_, indices) => indices}.distinct
+    val originalPairedBlocks = matchIndices.collect(processorBlocks)
+
+    val uniqueBlocks = (processorBlocks ++ orphanBlocks).foldLeft(Seq.empty[Block]) { (unique, current) =>
+      if (!unique.exists(current.approximateEquals(_))) {
+        current +: unique
       } else {
-        None
+        unique
       }
-    }).flatten
+    }
 
-    // Remove redundant faces and remove duplicates
-    val joinedBlocks = (reconstructedBlocks map { case block @ Block(center, _) =>
-      Block(center, block.nonRedundantFaces)
-    }).filter { block => block.faces.nonEmpty } // Filters blocks that actually aren't adjacent at all,
-                                                // but shared a processor joint
+    val currentOrphanBlocks = uniqueBlocks.diff(originalPairedBlocks)
```

Comment:
Yep I agree. But, playing devil's advocate for a moment: premature optimization can get us into trouble.